src/NHibernate/Linq/NhLinqExpression.cs

```diff
@@ -87,6 +87,9 @@ public IASTNode Translate(ISessionFactoryImplementor sessionFactory, bool filter
 
 			var requiredHqlParameters = new List<NamedParameterDescriptor>();
 			var queryModel = NhRelinqQueryParser.Parse(_expression);
+			// Use the modified TransparentIdentifierRemovingExpressionVisitor in order to support expressions that were created
+			// programmatically and do not mimic what the C# compiler generates.
+			queryModel.TransformExpressions(TransparentIdentifierRemovingExpressionVisitor.ReplaceTransparentIdentifiers);
```

Comment:
Fortunately that is cached in the query plan cache, so it is acceptable to traverse it again I think.

Reply:
I agree, the only thing that we could do is to make the capacity of the query cache configurable as currently is hardcoded to `128`
https://github.com/nhibernate/nhibernate-core/blob/7468a00a2207b411bddf5a9f2fe924d31bdfb378/src/NHibernate/Engine/Query/QueryPlanCache.cs#L29

which may be a little to small for larger applications. Of course, this should be done in a separate PR.

Reply:
> Of course, this should be done in a separate PR.

Especially since that could be a deep rabbit hole. I think we should instead get rid of this `SoftLimitMRUCache` and use something like `System.Runtime.MemoryCache`.

The limit is hardcoded to 128, but that is not a strong limit. `SoftLimitMRUCache` uses only weak references past 128 entries and lets the garbage collector decides whether to free them or not.